Framework/Kernel/src/StringTokenizer.cpp

```diff
@@ -0,0 +1,171 @@
+                                const std::string &delims) {
+  std::vector<std::string> output;
+  for_each_token(str.cbegin(), str.cend(), delims.cbegin(), delims.cend(),
+                 [&output](std::string::const_iterator first,
+                           std::string::const_iterator second) {
+                   output.emplace_back(first, second);
+                 });
+  return output;
+}
+
+// split keeping whitespace and ignoring empty tokens.
+std::vector<std::string> split1(const std::string &str,
+                                const std::string &delims) {
+  std::vector<std::string> output;
+  for_each_token(str.cbegin(), str.cend(), delims.cbegin(), delims.cend(),
+                 [&output](std::string::const_iterator first,
+                           std::string::const_iterator second) {
+                   if (first != second)
+                     output.emplace_back(first, second);
+                 });
+  return output;
+}
+
+// split trimming whitespace and keeping empty tokens.
+std::vector<std::string> split2(const std::string &str,
+                                const std::string &delims) {
+  std::vector<std::string> output;
+  for_each_token(str.cbegin(), str.cend(), delims.cbegin(), delims.cend(),
+                 [&output](std::string::const_iterator first,
+                           std::string::const_iterator second) {
+                   output.emplace_back(first, second);
+                   trim(output.back());
+                 });
+  return output;
+}
+
+// split trimming whitespace and ignoring empty tokens.
+std::vector<std::string> split3(const std::string &str,
+                                const std::string &delims) {
+  std::vector<std::string> output;
+  for_each_token(str.cbegin(), str.cend(), delims.cbegin(), delims.cend(),
+                 [&output](std::string::const_iterator first,
+                           std::string::const_iterator second) {
+                   if (first != second) {
+                     output.emplace_back(first, second);
+                     trim(output.back());
+                     if (output.back().empty())
+                       output.pop_back();
+                   }
+                 });
+  return output;
+}
+}
+
+/**
+ * Constructor requiring a string to tokenize and a string of separators.
+ * @param str Input string to be separated into tokens.
+ * @param separators List of characters used to separate the input string.
+ * @param options  tokenizer settings. The number can be found using the
+ * StringTokenizer::Options enum
+ * @throw Throws std::runtime_error if options > 7.
+ * @return a const reference to the index'th token.
+ */
+Mantid::Kernel::StringTokenizer::StringTokenizer(const std::string &str,
+                                                 const std::string &separators,
+                                                 unsigned options) {
+  // check options variable is in the range 0-7.
+  if (options > 7)
```

Comment:
One cannot throw an exception from inside the switch statement. I can, however move the check to after the switch so the check isn't performed in most cases.

I looked into changing to scoped enums, but every idea I can come up with is more complex. Specifically, combining options would be a challenge. Passing a container of enums or having a separate class to manage options takes more lines of code an requires more changes throughout the codebase.